File: agent-backend/app/agent/clarification.py
"""LLM-driven clarification before deep PTM research."""

from __future__ import annotations

import asyncio
import json
import logging
import re
from typing import Any

from app.agent.llm_fallback import log_llm_fallback
from app.agent.memory import InvestigationMemory

logger = logging.getLogger(__name__)

_CLARIFICATION_ASSESS_TIMEOUT_S = 12.0
_CLARIFICATION_MODES = frozenset({"research", "followup", "compare", "literature"})

_ASSESS_SYSTEM = """You are a PTM (post-translational modification) research planner for qPTM Agent.

Your job is NOT to decide whether a database lookup is possible — almost any gene+site can be \
looked up. Your job is to decide whether **one short clarification** would help the agent run a \
**deeper, more meaningful** investigation instead of a shallow generic summary.

## Default stance
When the user only names a protein/site/PTM (or gives a very short phrase) without stating \
**what they want to learn**, you SHOULD ask for clarification. The user can skip and run anyway.

## Ask for clarification when (common cases)
- Bare site mention: e.g. "TP53 S15磷酸化", "AKT1 S473", "STAT3 Y705 phosphorylation" — \
  no investigative angle stated
- Multiple equally plausible deep paths (WHO kinases / WHEN conditions / WHERE localization / \
  WHY function-disease) and the user did not prioritize
- Missing scope that would materially change retrieval: organism, stimulus, disease/cell line, \
  time point, comparison baseline
- Incomplete or truncated question; user has not finished stating the research goal

## Skip clarification ONLY when
- The user already states a clear investigative angle or specific factual question, e.g.:
  - "Which kinases phosphorylate TP53 S15?"
  - "TP53 S15 fold change under DNA damage"
  - "Functional role of TP53 S15 phosphorylation in apoptosis"
- Greeting, capability, or broad conceptual overview (not site research)
- Follow-up in the same session that already scoped the investigation

When asking, propose 1–2 **research-direction** fields (not just gene/site — those are often \
already known). Always include a free-text line for the user to state their goal in their own words.

Return ONLY valid JSON (no markdown fences):
{
  "needs_clarification": true or false,
  "intro": "one sentence in the user's language",
  "fields": [
    {
      "id": "snake_case",
      "label": "field label",
      "options": ["chip option 1", "chip option 2"],
      "option_values": ["optional", "parallel", "ids"],
      "allow_custom": true,
      "placeholder": "optional custom input hint"
    }
  ],
  "free_text": {
    "label": "bottom text area label",
    "placeholder": "hint for free-form input"
  },
  "submit_label": "button label",
  "skip_label": "skip button label"
}

If needs_clarification is false, return only: {"needs_clarification": false}
Max 2 fields. Match Chinese if the user wrote Chinese, otherwise English."""


def _detect_lang(question: str) -> str:
    zh = sum(1 for ch in question if "\u4e00" <= ch <= "\u9fff")
    latin = sum(1 for ch in question if "a" <= ch.lower() <= "z")
    return "zh" if zh >= 2 and zh >= latin * 0.35 else "en"


def should_assess_clarification(mode: str) -> bool:
    return (mode or "").lower() in _CLARIFICATION_MODES


def _parse_json_object(raw: str) -> dict[str, Any] | None:
    text = (raw or "").strip()
    if not text:
        return None
    if text.startswith("```"):
        text = re.sub(r"^```(?:json)?\s*", "", text)
        text = re.sub(r"\s*```$", "", text)
    try:
        data = json.loads(text)
    except json.JSONDecodeError:
        return None
    return data if isinstance(data, dict) else None


def _normalize_field(field: Any) -> dict[str, Any] | None:
    if not isinstance(field, dict):
        return None
    field_id = str(field.get("id") or "").strip()
    label = str(field.get("label") or "").strip()
    if not field_id or not label:
        return None
    options = field.get("options") or []
    if not isinstance(options, list):
        options = []
    options = [str(o).strip() for o in options if str(o).strip()][:6]
    option_values = field.get("option_values") or []
    if not isinstance(option_values, list):
        option_values = []
    option_values = [str(v).strip() for v in option_values][: len(options)]
    return {
        "id": field_id,
        "label": label,
        "options": options,
        "option_values": option_values,
        "allow_custom": bool(field.get("allow_custom", True)),
        "placeholder": str(field.get("placeholder") or ""),
    }


def _normalize_clarification_payload(
    data: dict[str, Any],
    question: str,
    lang: str,
) -> dict[str, Any] | None:
    if not data.get("needs_clarification"):
        return None

    intro = str(data.get("intro") or "").strip()
    if not intro:
        intro = (
            "为进行更有深度的位点级研究，请补充以下信息（也可跳过直接执行）："
            if lang == "zh"
            else "To run a deeper site-level investigation, please add the following (or skip and run anyway):"
        )

    fields: list[dict[str, Any]] = []
    for raw_field in (data.get("fields") or [])[:2]:
        normalized = _normalize_field(raw_field)
        if normalized:
            fields.append(normalized)

    free_raw = data.get("free_text") or {}
    if not isinstance(free_raw, dict):
        free_raw = {}
    free_label = str(free_raw.get("label") or "").strip()
    free_placeholder = str(free_raw.get("placeholder") or "").strip()
    if not free_label:
        free_label = "其他补充（可选）" if lang == "zh" else "Additional notes (optional)"
    if not free_placeholder:
        free_placeholder = (
            "请用一句话说明您最想了解的内容…"
            if lang == "zh"
            else "Describe what you most want to learn…"
        )

    submit_label = str(data.get("submit_label") or "").strip()
    skip_label = str(data.get("skip_label") or "").strip()
    if not submit_label:
        submit_label = "开始研究" if lang == "zh" else "Start research"
    if not skip_label:
        skip_label = "跳过，直接执行" if lang == "zh" else "Skip and run"

    # Require at least one actionable input (chips or free text)
    if not fields and not free_label:
        return None

    return {
        "intro": intro,
        "original_question": question,
        "fields": fields,
        "free_text": {
            "id": "notes",
            "label": free_label,
            "placeholder": free_placeholder,
        },
        "submit_label": submit_label,
        "skip_label": skip_label,
    }


def _build_assess_prompt(
    question: str,
    entities: dict[str, Any],
    memory: InvestigationMemory,
    history: list[dict[str, str]],
) -> str:
    entity_lines = [
        f"- gene: {entities.get('gene') or 'unknown'}",
        f"- site/position: {entities.get('position') or 'unknown'}",
        f"- ptm_type: {entities.get('ptm_type') or 'unknown'}",
        f"- organism: {entities.get('organism') or memory.organism or 'unknown'}",
        f"- pmid: {entities.get('pmid') or 'none'}",
    ]
    hist = ""
    if history:
        tail = history[-4:]
        hist = "\n".join(f"{m.get('role', 'user')}: {m.get('content', '')[:300]}" for m in tail)
    memory_block = memory.to_prompt_block() or "(empty)"
    q = question.strip()
    wordish = len(q.split()) + len(re.findall(r"[\u4e00-\u9fff]", q))
    return (
        f"## Parsed entities\n" + "\n".join(entity_lines) + "\n\n"
        f"## Session memory\n{memory_block}\n\n"
        f"## Recent conversation\n{hist or '(none)'}\n\n"
        f"## Question brevity signal (hint only — you decide)\n"
        f"- characters: {len(q)}\n"
        f"- approximate tokens/words: {wordish}\n"
        f"- states investigative angle explicitly: unknown — use your judgment\n\n"
        f"## Current user question\n{question}\n"
    )


async def assess_clarification_need(
    llm: Any,
    question: str,
    entities: dict[str, Any],
    memory: InvestigationMemory,
    history: list[dict[str, str]],
    *,
    lang: str | None = None,
) -> dict[str, Any] | None:
    """Ask the LLM whether to show a clarification modal; return payload or None."""
    lang = lang or _detect_lang(question)
    prompt = _build_assess_prompt(question, entities, memory, history)

    def _call() -> str:
        result = llm.chat_completion(
            messages=[
                {"role": "system", "content": _ASSESS_SYSTEM},
                {"role": "user", "content": prompt},
            ],
            temperature=0.1,
            max_tokens=800,
            disable_thinking=True,
        )
        return result.get("content") or ""

    try:
        raw = await asyncio.wait_for(asyncio.to_thread(_call), timeout=_CLARIFICATION_ASSESS_TIMEOUT_S)
    except Exception as exc:
        log_llm_fallback(
            "Clarification assessment",
            exc,
            timeout_s=_CLARIFICATION_ASSESS_TIMEOUT_S,
            detail="skipping clarification modal",
        )
        return None

    data = _parse_json_object(raw)
    if not data:
        logger.warning("Clarification assessment returned non-JSON: %s", raw[:200])
        return None

    payload = _normalize_clarification_payload(data, question, lang)
    if payload:
        logger.info("Clarification requested by LLM for: %s", question[:80])
    return payload


def build_literature_clarification_request(
    question: str,
    *,
    lang: str | None = None,
) -> dict[str, Any]:
    """Deterministic clarification when the user wants papers but no target is set."""
    lang = lang or _detect_lang(question)
    if lang == "zh":
        return {
            "intro": "为了推荐有针对性的文献，请补充您关注的蛋白、修饰位点或研究主题（也可跳过，系统将尽量检索）：",
            "original_question": question,
            "fields": [
                {
                    "id": "literature_focus",
                    "label": "您更关注哪类文献？",
                    "options": [
                        "位点机制与功能",
                        "上游激酶/调控",
                        "疾病与临床关联",
                        "实验/组学方法",
                        "领域综述",
                    ],
                    "option_values": [
                        "mechanism",
                        "kinase",
                        "disease",
                        "methods",
                        "review",
                    ],
                    "allow_custom": True,
                    "placeholder": "其他方向…",
                },
            ],
            "free_text": {
                "id": "notes",
                "label": "蛋白、位点或研究主题",
                "placeholder": "例如：TP53 S15 磷酸化、AKT1 S473、DNA 损伤应激下的 p53 修饰…",
            },
            "submit_label": "推荐文献",
            "skip_label": "跳过，直接检索",
        }
    return {
        "intro": (
            "To recommend relevant papers, please specify the protein/site or research topic "
            "(or skip and I will search with what you provided):"
        ),
        "original_question": question,
        "fields": [
            {
                "id": "literature_focus",
                "label": "What type of papers do you want?",
                "options": [
                    "Site mechanism & function",
                    "Upstream kinases/regulation",
                    "Disease & clinical links",
                    "Methods / omics",
                    "Field review",
                ],
                "option_values": ["mechanism", "kinase", "disease", "methods", "review"],
                "allow_custom": True,
                "placeholder": "Other focus…",
            },
        ],
        "free_text": {
            "id": "notes",
            "label": "Protein, site, or topic",
            "placeholder": "e.g. TP53 S15 phosphorylation, AKT1 S473, p53 under DNA damage…",
        },
        "submit_label": "Recommend papers",
        "skip_label": "Skip and search",
    }


def merge_clarification_answers(
    original_question: str,
    selections: dict[str, str],
    free_text: str = "",
) -> str:
    """Merge modal answers into an enriched research question."""
    notes = (free_text or "").strip()
    sel_parts = [v.strip() for v in selections.values() if v and str(v).strip()]

    if notes:
        if len(notes) >= max(20, len(original_question) * 0.5) or "?" in notes or "？" in notes:
            base = notes
        else:
            zh = bool(re.search(r"[\u4e00-\u9fff]", original_question + notes))
            base = f"{original_question.strip()}\n{'补充' if zh else 'Notes'}: {notes}"
    else:
        base = original_question.strip()

    if sel_parts:
        zh = bool(re.search(r"[\u4e00-\u9fff]", base))
        joined = "；".join(sel_parts) if zh else "; ".join(sel_parts)
        base = f"{base}\n{'研究侧重' if zh else 'Focus'}: {joined}"

    return base

File: agent-runtime/src/agent/clarification.ts
import { detectLang } from "./gate.js";
import {
  InvestigationMemory,
  extractPositionFromText,
  memoryPromptBlock,
  mergeEntities,
  parseEntities,
} from "../context/memory.js";
import { getLlm } from "../llm/client.js";
import type { AgentEvent } from "../sse.js";

export interface ClarificationOption {
  label: string;
  description?: string;
}

export interface ClarificationField {
  id: string;
  label: string;
  /** Short question under the field title. */
  prompt?: string;
  options: ClarificationOption[];
  allow_custom?: boolean;
  placeholder?: string;
}

export interface ClarificationPayload {
  needs_clarification: boolean;
  intro?: string;
  fields?: ClarificationField[];
  free_text?: { label: string; placeholder: string };
  submit_label?: string;
  skip_label?: string;
}

function opt(label: string, description = ""): ClarificationOption {
  return { label, description };
}

/** Minimal fallback only when LLM fails — still question-shaped, not a fixed WHO/WHEN template. */
function fallbackClarification(message: string, memory: InvestigationMemory): ClarificationPayload {
  const lang = detectLang(message);
  const target = [memory.gene, memory.position ? `S${memory.position}` : "", memory.ptm_type]
    .filter(Boolean)
    .join(" ")
    .trim();
  const subject = target || (lang === "zh" ? "该 PTM 问题" : "this PTM question");

  if (lang === "zh") {
    return {
      needs_clarification: true,
      intro: `在深入调研「${subject}」前，请先确认最关键的一点（可跳过）：`,
      fields: [
        {
          id: "priority",
          label: "本次优先弄清什么？",
          prompt: "选一项最贴近你当前目标的方向，或在下方手输。",
          options: [
            opt("上游如何调控", "激酶、酶或刺激如何修饰该位点"),
            opt("何时 / 何种条件变化", "定量动力学、处理条件与比较"),
            opt("功能与疾病意义", "功能后果、稳定性或疾病关联"),
            opt("文献机制", "已有机制链条与争议"),
          ],
          allow_custom: true,
          placeholder: "也可直接写你的目标…",
        },
      ],
      free_text: {
        label: "补充说明（可选）",
        placeholder: "物种/细胞系、比较条件、疾病背景等",
      },
      submit_label: "开始深度调研",
      skip_label: "跳过，直接调研",
    };
  }

  return {
    needs_clarification: true,
    intro: `Before deep research on “${subject}”, confirm the main priority (optional — you can skip):`,
    fields: [
      {
        id: "priority",
        label: "What should we prioritize?",
        prompt: "Pick the closest goal, or type your own below.",
        options: [
          opt("Upstream regulation", "Kinases/enzymes/stimuli acting on the site"),
          opt("Conditions & dynamics", "Quantitative changes and experimental conditions"),
          opt("Function & disease", "Functional consequences, stability, disease links"),
          opt("Literature mechanisms", "Known mechanistic chains and debates"),
        ],
        allow_custom: true,
        placeholder: "Or type your goal…",
      },
    ],
    free_text: {
      label: "Additional context (optional)",
      placeholder: "Organism/cell line, comparison, disease background…",
    },
    submit_label: "Start deep research",
    skip_label: "Skip and research",
  };
}

function normalizeOption(raw: unknown): ClarificationOption | null {
  if (typeof raw === "string" && raw.trim()) return { label: raw.trim() };
  if (!raw || typeof raw !== "object") return null;
  const o = raw as Record<string, unknown>;
  const label = String(o.label || o.value || "").trim();
  if (!label) return null;
  return {
    label,
    description: String(o.description || o.desc || "").trim() || undefined,
  };
}

function normalizePayload(raw: unknown, message: string, memory: InvestigationMemory): ClarificationPayload {
  if (!raw || typeof raw !== "object") return fallbackClarification(message, memory);
  const data = raw as Record<string, unknown>;

  if (data.needs_clarification === false) {
    return { needs_clarification: false };
  }

  const lang = detectLang(message);
  const fieldsIn = Array.isArray(data.fields) ? data.fields : [];
  const fields: ClarificationField[] = [];

  for (let i = 0; i < fieldsIn.length && fields.length < 3; i += 1) {
    const f = fieldsIn[i];
    if (!f || typeof f !== "object") continue;
    const fr = f as Record<string, unknown>;
    const options = (Array.isArray(fr.options) ? fr.options : [])
      .map(normalizeOption)
      .filter((x): x is ClarificationOption => Boolean(x))
      .slice(0, 6);
    if (options.length < 2) continue;
    const id = String(fr.id || `q${fields.length + 1}`).replace(/[^\w-]/g, "_").slice(0, 40) || `q${fields.length + 1}`;
    fields.push({
      id,
      label: String(fr.label || (lang === "zh" ? `问题 ${fields.length + 1}` : `Question ${fields.length + 1}`)).slice(0, 80),
      prompt: String(fr.prompt || fr.question || "").slice(0, 200) || undefined,
      options,
      allow_custom: fr.allow_custom !== false,
      placeholder: String(fr.placeholder || (lang === "zh" ? "其他：直接输入…" : "Other: type here…")).slice(0, 160),
    });
  }

  if (!fields.length) return fallbackClarification(message, memory);

  const free = (data.free_text && typeof data.free_text === "object"
    ? (data.free_text as Record<string, unknown>)
    : null);

  return {
    needs_clarification: true,
    intro: String(
      data.intro ||
        (lang === "zh"
          ? "开始深度调研前，我想先确认以下几点（均可跳过）："
          : "Before deep research, I’d like to confirm a few points (all optional):"),
    ).slice(0, 300),
    fields,
    free_text: {
      label: String(free?.label || (lang === "zh" ? "补充说明（可选）" : "Additional context (optional)")).slice(0, 80),
      placeholder: String(
        free?.placeholder ||
          (lang === "zh"
            ? "物种/细胞系、比较条件、疾病背景等"
            : "Organism/cell line, comparison, disease background…"),
      ).slice(0, 160),
    },
    submit_label: String(data.submit_label || (lang === "zh" ? "开始深度调研" : "Start deep research")).slice(0, 40),
    skip_label: String(data.skip_label || (lang === "zh" ? "跳过，直接调研" : "Skip and research")).slice(0, 40),
  };
}

export interface ClarifyRoundOpts {
  /** Completed clarification rounds so far (0 = first ask). */
  round?: number;
  maxRounds?: number;
}

/**
 * Agent-generated clarification: judge what is still ambiguous,
 * then propose 1–3 targeted questions — may run across multiple rounds.
 */
export async function buildDeepResearchClarification(
  message: string,
  memory?: InvestigationMemory,
  roundOpts: ClarifyRoundOpts = {},
): Promise<ClarificationPayload> {
  const mem = memory || ({ ...parseEntities(message) } as InvestigationMemory);
  const lang = detectLang(message);
  const round = roundOpts.round ?? 0;
  const maxRounds = roundOpts.maxRounds ?? 4;

  const system =
    lang === "zh"
      ? `你是 qPTM 深度调研助手。请判断：以当前信息，是否还需要向用户澄清后才能做更精准的调研。
规则：
1. 允许多轮澄清。已完成 ${round}/${maxRounds} 轮。只问**仍然不明确且对本次调研关键**的点；不要重复用户已回答过的内容。
2. 不要套固定 WHO/WHEN/WHERE/WHY 模板；每次 1–3 个字段，每字段 2–5 个贴合选项（含简短 description）。
3. 若蛋白/基因已知但无具体残基编号，且目标是位点级（调控网络、上游激酶、功能等），应问 id=site，选项写清残基（如「S18（小鼠）/ S15（人）」）。若用户只选了「单个位点」却没给编号，下一轮应追问具体位点。
4. 信息已足够，或用户已明确可先按假设推进时，返回 needs_clarification=false（调研阶段仍可在报告里标注假设或再提问）。
5. 不要为了凑问题而提问；不关键就跳过。
6. 字段 id 用英文 snake_case；文案用中文。只输出 JSON，不要 markdown。

JSON 格式：
{
  "needs_clarification": true|false,
  "intro": "简短说明",
  "fields": [
    {
      "id": "priority",
      "label": "标题",
      "prompt": "一句问句",
      "options": [{"label":"...","description":"..."}],
      "allow_custom": true,
      "placeholder": "其他…"
    }
  ],
  "free_text": {"label":"补充说明（可选）","placeholder":"..."},
  "submit_label": "开始深度调研",
  "skip_label": "跳过，直接调研"
}`
      : `You are the qPTM deep-research assistant. Decide whether more clarification is still needed.
Rules:
1. Multi-round clarification is allowed. Completed ${round}/${maxRounds} rounds. Ask ONLY what is still ambiguous and material; do not re-ask answered points.
2. No fixed WHO/WHEN/WHERE/WHY template; 1–3 fields, 2–5 tailored options each.
3. If gene/protein is known but residue is missing for a site-level goal, ask id=site with concrete residues. If the user chose “single site” without a number, follow up on the residue.
4. If information is sufficient (or the user can proceed under a stated assumption), return needs_clarification=false.
5. Do not invent questions. Output JSON only.

JSON schema:
{
  "needs_clarification": true|false,
  "intro": "...",
  "fields": [{"id":"...","label":"...","prompt":"...","options":[{"label":"...","description":"..."}],"allow_custom":true,"placeholder":"..."}],
  "free_text": {"label":"...","placeholder":"..."},
  "submit_label": "Start deep research",
  "skip_label": "Skip and research"
}`;

  const user = [
    `User question / accumulated context:\n${message}`,
    memoryPromptBlock(mem),
    `Parsed entities: gene=${mem.gene || ""} position=${mem.position || ""} ptm=${mem.ptm_type || ""} organism=${mem.organism || ""}`,
    `Clarification round: ${round} (0=first ask). Max rounds: ${maxRounds}.`,
  ].join("\n");

  try {
    const llm = getLlm();
    const { content } = await llm.chatCompletion(
      [
        { role: "system", content: system },
        { role: "user", content: user },
      ],
      { maxTokens: 1200, temperature: 0.3 },
    );
    const text = content.replace(/^```(?:json)?\s*/i, "").replace(/\s*```$/i, "").trim();
    const parsed = JSON.parse(text) as unknown;
    return normalizePayload(parsed, message, mem);
  } catch {
    // On follow-up rounds, fail open (start research) rather than forcing a generic card again.
    if (round > 0) return { needs_clarification: false };
    return fallbackClarification(message, mem);
  }
}

export function mergeClarification(
  original: string,
  selections: Record<string, string>,
  freeText: string,
): string {
  const parts = [original];
  for (const [k, v] of Object.entries(selections)) {
    if (!v) continue;
    parts.push(`${k}: ${v}`);
  }
  if (freeText.trim()) parts.push(freeText.trim());
  return parts.join("\n");
}

/** Fold clarification choices into memory (site labels, organism, etc.). */
export function applyClarificationToMemory(
  memory: InvestigationMemory,
  selections: Record<string, string>,
  freeText: string,
): void {
  const texts = [...Object.values(selections), freeText].filter(Boolean);
  for (const t of texts) {
    mergeEntities(memory, parseEntities(t), t);
    const pos = extractPositionFromText(t);
    if (pos && !memory.position) memory.position = pos;
    if (/小鼠|mouse/i.test(t)) memory.organism = "mouse";
    if (/人源|人类|human/i.test(t) && memory.organism !== "mouse") memory.organism = "human";
  }
}

export function clarificationEvent(payload: ClarificationPayload): AgentEvent {
  return { type: "clarification_request", ...payload };
}
